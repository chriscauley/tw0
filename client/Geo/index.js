const { range } = require('lodash')
const alpha = 'abcdefghijklmnopqrstuvwxyz'
const ALPHA = alpha.toUpperCase()
const numbers = '0123456789'
export const alphabet = alpha.split('')
export const alphanum = (alpha+ALPHA+numbers).split('')
export const numalpha = (numbers+alpha+ALPHA).split('')

const mod = (n, d) => ((n % d) + d) % d

const Look = (geo) => {
  const make = (shape, dist) => {
    const R = 1
    if (look[shape][R][dist]) {
      return
    }
    if (dist > 0 && !look[shape][R][dist - 1]) {
      make(shape, dist - 1)
    }
    if (shape.startsWith('__')) {
      // dunder means "outer shell of look"
      geo.dindexes.forEach(dindex => {
        look[shape][dindex][dist] = shapes[shape.slice(2)](dist)
      })
    } else {
      if (!look['__' + shape][R][dist]) {
        // need outer shells to make filled looks
        make('__' + shape, dist)
      }
      geo.dindexes.forEach(dindex => {
        look[shape][dindex][dist] = []
        range(dist + 1).forEach((_dist) => {
          look[shape][dindex][dist] = look[shape][dindex][dist].concat(look['__' + shape][dindex][_dist])
        })
      })
    }
  }

  const shapes = {
    box: (dist) => {
      const out = []
      geo.dindexes.forEach((dindex, i) => {
        const o_dindex = geo.dindexes[(i + 1) % geo.dindexes.length] // orthogonal direction
        range(dist * 2 + 1).forEach((o_dist) => {
          out.push(dist * dindex + o_dist * o_dindex)
        })
        out.pop() // last corner is repeated
      })
      return out
    },
    circle: (dist) => {
      const out = []
      geo.dindexes.forEach((dindex, i) => {
        out.push(dindex * dist)
        const o_dindex = geo.dindexes[(i + 1) % geo.dindexes.length] // orthogonal direction
        range(1, dist).forEach((o_dist) =>
          out.push((dist - o_dist) * dindex + o_dist * o_dindex),
        )
      })
      return out
    },
  }

  const look = (shape, index, dist, dindex) => {
    if (!look[shape][dindex][dist]) {
      make(shape, dist)
    }
    return look[shape][dindex][dist].map((dindex) => index + dindex)
  }

  Object.keys(shapes).forEach((shape) => {
    look[shape] = {}
    look['__' + shape] = {}
    geo.dindexes.forEach(dindex => {
      look[shape][dindex] = [[]] // all geometries only see nothing at range 0
      look['__' + shape][dindex] = [[]]
    })
    make(shape, 1)
  })

  look.inBounds = (shape, index, dist) => {
    // like look, but won't cross out of bounds
    const x0 = geo.i2xy(index)[0]
    const xys = look(shape, index, dist)
      .map((i) => geo.i2xy(i))
      .filter(
        (xy) =>
          Math.abs(xy[0] - x0) <= dist &&
          xy[1] < geo.H &&
          xy[1] >= geo.y0 &&
          xy[1] < geo.y0 + geo.H,
      )
    return xys.map((xy) => geo.xy2i(xy))
  }

  return look
}

const Geo = (x0, x_max, y0, y_max) => {
  if (y0 === undefined) {
    y_max = x_max - 1
    x_max = x0 - 1
    x0 = 0
    y0 = 0
  }

  const W = Math.abs(x_max - x0) + 1
  const H = Math.abs(y_max - y0) + 1

  const geo = {
    x0,
    y0,
    W,
    H,
    xys: [],
    indexes: [],
    AREA: W * H,
    dindexes: [-W, -1, 1, W ], // u, l, r, d
    rot_dindexes: {
      [-W]: [-W, -1, 1, W ], // u, l, r, d
      [W]: [W, 1, -1, -W ], // d, l, r, u
      [1]: [1, -W, W, -1], // r, u, d, l
      [-1]: [-1, W, -W, 1], // l, d, u, r
    },
    i2xy: (i) => [mod(i, W), Math.floor(i / W)],
    xy2i: (xy) => xy[0] + xy[1] * W,
    print(board, {from_xy=[x0,y0], to_xy=[x_max, y_max], delimiter='',empty=' ', extras, title}={}) {
      const xs = range(from_xy[0], to_xy[0]+1)
      const ys = range(from_xy[1], to_xy[1]+1)
      const lines = ys.map((y) => (
        xs.map((x) => board[this.xy2i([x, y])])
          .map(s => s=== undefined ? empty : s)
          .join(delimiter)
      ))
      if (extras) {
        extras.forEach((e, i) => lines[i].push('  ' + e))
      }
      title && lines.unshift(title)
      return lines.join('\n')
    },
    log: (board, options) => console.log(geo.print(board, options)),
    makeBoard(xys, values = []) {
      const board = {}
      xys.forEach((xy, i) => {
        board[this.xy2i(xy)] = values[i] || i
      })
      return board
    },
    inBounds(xy) {
      return xy[0] >= x0 && xy[0] < x0 + W && xy[1] >= y0 && xy[1] < y0 + H
    },
    eachXY(f) {
      console.warn('eachXY is depracated, use geo.xys.forEach instead')
      range(x0, x_max+1).forEach(
        x => range(y0, y_max+1).forEach(
          y => f([x,y])
        )
      )
    },
    slice(board, xy, W, H, _default) {
      const out = []
      const ys = range(xy[1], xy[1] + H)
      range(xy[0], xy[0]+W).forEach(
        x => ys.forEach(
          y => {
            const v = board[geo.xy2i([x,y])]
            out.push(v === undefined ? _default : v)
          }
        )
      )
      return out
    },
  }

  geo.CENTER = geo.xy2i([Math.floor((geo.x0+geo.W)/2), Math.floor((geo.y0+geo.H)/2)])

  geo._wrapped_dindexes = geo.dindexes.concat(geo.dindexes)
  geo.dindexes.forEach((dindex, i) => {
    geo.rot_dindexes[dindex] = geo._wrapped_dindexes.slice(i, i+geo.dindexes.length)
  })

  range(x0, x_max+1).forEach(
    x => range(y0, y_max+1).forEach(
      y => {
        const xy = [x,y]
        geo.xys.push()
        geo.indexes.push(geo.xy2i(xy))
      }
    )
  )

  geo.look = Look(geo)
  return geo
}
const assert = (bool, exception) => {
  if (!bool) {
    throw exception
  }
}

const log = (...args) => {
  if (log.isOn()) {
    console.log(...args)
  }
}

log.isOn = () => process.argv.includes('-v')

const answer = (text, value, expected) => {
  if (expected !== undefined) {
    assert(value === expected, `Bad answer for ${text}, ${value} !== ${expected}`)
  }
  console.log(text, value)
}

// eslint-disable-next-line
export default Geo
//   numalpha,
//   alphanum,
//   alphabet: alphanum.slice(0,26),
//   Geo,
//   assert,
//   log,
//   answer,
//   mod,
// }
