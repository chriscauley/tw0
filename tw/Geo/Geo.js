import { range } from 'lodash'

import { mod } from '../utils'
import Look from './Look'

const geo_cache = {}

export default (x0, x_max, y0, y_max) => {
  const key = `${x0},${x_max},${y0},${y_max}`
  if (geo_cache[key]) {
    return geo_cache[key]
  }
  if (x_max === undefined) {
    x_max = x0
  }
  if (y0 === undefined) {
    y_max = x_max - 1
    x_max = x0 - 1
    x0 = 0
    y0 = 0
  }

  const W = Math.abs(x_max - x0) + 1
  const H = Math.abs(y_max - y0) + 1

  const geo = (geo_cache[key] = {
    x0,
    y0,
    W,
    H,
    xys: [],
    indexes: [],
    AREA: W * H,
    _dindex_names: ['u', 'l', 'r', 'd'],
    _dindex_chars: ['^', '<', '>', 'v'],
    _dindex2char: {},
    _name2dindex: {},
    _dindex2name: {},
    dindexes: [-W, -1, 1, W], // u, l, r, d
    rot_dindexes: {
      [-W]: [-W, -1, 1, W], // u, l, r, d
      [W]: [W, 1, -1, -W], // d, l, r, u
      [1]: [1, -W, W, -1], // r, u, d, l
      [-1]: [-1, W, -W, 1], // l, d, u, r
    },
    index2xy: (i) => [mod(i, W), Math.floor(i / W)],
    xy2index: (xy) => mod(xy[0] + xy[1] * W, geo.AREA),
    print(
      board,
      {
        from_xy = [x0, y0],
        to_xy = [x_max, y_max],
        delimiter = '',
        empty = ' ',
        extras,
        title,
      } = {},
    ) {
      const xs = range(from_xy[0], to_xy[0] + 1)
      const ys = range(from_xy[1], to_xy[1] + 1)
      const lines = ys.map((y) =>
        xs
          .map((x) => board[this.xy2index([x, y])])
          .map((s) => (s === undefined ? empty : s))
          .join(delimiter),
      )
      if (extras) {
        extras.forEach((e, i) => (lines[i] += '\t' + e))
      }
      title && lines.unshift(title)
      return lines.join('\n')
    },
    inBounds(xy) {
      return xy[0] >= x0 && xy[0] < x0 + W && xy[1] >= y0 && xy[1] < y0 + H
    },
    slice(xy, W, H) {
      const out = []
      const ys = range(xy[1], xy[1] + H)
      const xs = range(xy[0], xy[0] + W)
      ys.forEach((y) => xs.forEach((x) => out.push(geo.xy2index([x, y]))))
      return out
    },
    floorDindex(dindex) {
      // if dindex is a multiple of W it is in thy y direction
      // otherwise it is in the x direction
      return (dindex % W === 0 ? W : 1) * Math.sign(dindex)
    },
  })

  geo.dindexes.forEach((dindex, i) => {
    const name = geo._dindex_names[i]
    geo._dindex2name[dindex] = name
    geo._name2dindex[name] = dindex
    geo._dindex2char[dindex] = geo._dindex_chars[i]
  })

  geo.CENTER = geo.xy2index([Math.floor((geo.x0 + geo.W) / 2), Math.floor((geo.y0 + geo.H) / 2)])

  range(y0, y_max + 1).forEach((y) =>
    range(x0, x_max + 1).forEach((x) => {
      const xy = [x, y]
      geo.xys.push(xy)
      geo.indexes.push(geo.xy2index(xy))
    }),
  )

  geo.look = Look(geo)
  return geo
}
