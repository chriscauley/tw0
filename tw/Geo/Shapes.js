import { range } from 'lodash'

const Shapes = (geo) => ({
  circle: (dist, dindex) => {
    const out = []
    const o_dindex = geo.rot_dindexes[dindex][1]
    out.push(dindex * dist)
    range(1, dist + 1).forEach((i) => {
      out.push(dindex * (dist - i) - o_dindex * i)
      out.push(dindex * (dist - i) + o_dindex * i)
    })
    range(1, dist).forEach((i) => {
      out.push(-dindex * (dist - i) - o_dindex * i)
      out.push(-dindex * (dist - i) + o_dindex * i)
    })
    out.push(dindex * -dist)
    return out
  },
  box: (dist, dindex) => {
    const out = []
    const o_dindex = geo.rot_dindexes[dindex][1]

    // top row
    range(-dist, dist + 1).forEach((i) => out.push(dist * dindex - i * o_dindex))

    // left and right sides in middle
    range(1, 2 * dist).forEach((i) => {
      out.push((dist - i) * dindex + o_dindex * dist)
      out.push((dist - i) * dindex - o_dindex * dist)
    })

    // bottom row
    range(-dist, dist + 1).forEach((i) => out.push(-dist * dindex - i * o_dindex))
    return out
  },
  three: (dist, dindex) => {
    const target = dist * dindex
    const o_dindex = geo.rot_dindexes[dindex][1]
    return [target, target + o_dindex, target - o_dindex]
  },
  cone: (dist, dindex) => {
    const target = dist * dindex
    const out = [target]
    const o_dindex = geo.rot_dindexes[dindex][1]
    range(1, dist).forEach((o_dist) => {
      out.push(target + o_dist * o_dindex)
      out.push(target - o_dist * o_dindex)
    })
    return out
  },
})

Shapes.list = []

Object.keys(Shapes()).forEach((shape) => {
  Shapes.list.push('__' + shape)
  Shapes.list.push(shape)
})

export default Shapes
