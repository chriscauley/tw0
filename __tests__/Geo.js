import Geo, {numalpha} from '../client/Geo'

test('xy2index', () => {
  const geo = Geo(5, 5)
  const b = {}
  const set = (xy, v) => b[geo.xy2index(xy)] = v
  set([2, 2], 1)
  set([-1,0], 2)
  set([3,-2], 3)
  set([1, 8], 4)
  set([5, 3], 5)
  set([6, 0], 6)
  expect(geo.print(b, {empty:'.'})).toMatchSnapshot()
})