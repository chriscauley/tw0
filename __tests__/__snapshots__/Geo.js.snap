// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`look.__box 1`] = `
"__box 1 u
012
3 4
567"
`;

exports[`look.__box 2`] = `
"__box 2 u
01234
5   6
7   8
9   a
bcdef"
`;

exports[`look.__box 3`] = `
"__box 3 u
0123456
7     8
9     a
b     c
d     e
f     g
hijklmn"
`;

exports[`look.__box 4`] = `
"__box 4 u
012345678
9       a
b       c
d       e
f       g
h       i
j       k
l       m
nopqrstuv"
`;

exports[`look.__box 5`] = `
"__box 1 l
247
1 6
035"
`;

exports[`look.__box 6`] = `
"__box 2 l
468af
3   e
2   d
1   c
0579b"
`;

exports[`look.__box 7`] = `
"__box 3 l
68acegn
5     m
4     l
3     k
2     j
1     i
079bdfh"
`;

exports[`look.__box 8`] = `
"__box 4 l
8acegikmv
7       u
6       t
5       s
4       r
3       q
2       p
1       o
09bdfhjln"
`;

exports[`look.__box 9`] = `
"__box 1 r
530
6 1
742"
`;

exports[`look.__box 10`] = `
"__box 2 r
b9750
c   1
d   2
e   3
fa864"
`;

exports[`look.__box 11`] = `
"__box 3 r
hfdb970
i     1
j     2
k     3
l     4
m     5
ngeca86"
`;

exports[`look.__box 12`] = `
"__box 4 r
nljhfdb90
o       1
p       2
q       3
r       4
s       5
t       6
u       7
vmkigeca8"
`;

exports[`look.__box 13`] = `
"__box 1 d
765
4 3
210"
`;

exports[`look.__box 14`] = `
"__box 2 d
fedcb
a   9
8   7
6   5
43210"
`;

exports[`look.__box 15`] = `
"__box 3 d
nmlkjih
g     f
e     d
c     b
a     9
8     7
6543210"
`;

exports[`look.__box 16`] = `
"__box 4 d
vutsrqpon
m       l
k       j
i       h
g       f
e       d
c       b
a       9
876543210"
`;

exports[`look.__circle 1`] = `
"__circle 1 u
 0 
2 1
 3 "
`;

exports[`look.__circle 2`] = `
"__circle 2 u
  0  
 2 1 
4   3
 6 5 
  7  "
`;

exports[`look.__circle 3`] = `
"__circle 3 u
   0   
  2 1  
 4   3 
6     5
 a   9 
  8 7  
   b   "
`;

exports[`look.__circle 4`] = `
"__circle 4 u
    0    
   2 1   
  4   3  
 6     5 
8       7
 e     d 
  c   b  
   a 9   
    f    "
`;

exports[`look.__circle 5`] = `
"__circle 1 l
 1 
0 3
 2 "
`;

exports[`look.__circle 6`] = `
"__circle 2 l
  3  
 1 5 
0   7
 2 6 
  4  "
`;

exports[`look.__circle 7`] = `
"__circle 3 l
   5   
  3 9  
 1   7 
0     b
 2   8 
  4 a  
   6   "
`;

exports[`look.__circle 8`] = `
"__circle 4 l
    7    
   5 d   
  3   b  
 1     9 
0       f
 2     a 
  4   c  
   6 e   
    8    "
`;

exports[`look.__circle 9`] = `
"__circle 1 r
 2 
3 0
 1 "
`;

exports[`look.__circle 10`] = `
"__circle 2 r
  4  
 6 2 
7   0
 5 1 
  3  "
`;

exports[`look.__circle 11`] = `
"__circle 3 r
   6   
  a 4  
 8   2 
b     0
 7   1 
  9 3  
   5   "
`;

exports[`look.__circle 12`] = `
"__circle 4 r
    8    
   e 6   
  c   4  
 a     2 
f       0
 9     1 
  b   3  
   d 5   
    7    "
`;

exports[`look.__circle 13`] = `
"__circle 1 d
 3 
1 2
 0 "
`;

exports[`look.__circle 14`] = `
"__circle 2 d
  7  
 5 6 
3   4
 1 2 
  0  "
`;

exports[`look.__circle 15`] = `
"__circle 3 d
   b   
  7 8  
 9   a 
5     6
 3   4 
  1 2  
   0   "
`;

exports[`look.__circle 16`] = `
"__circle 4 d
    f    
   9 a   
  b   c  
 d     e 
7       8
 5     6 
  3   4  
   1 2   
    0    "
`;

exports[`look.__cone 1`] = `
"__cone 1 u
 0 
   
   "
`;

exports[`look.__cone 2`] = `
"__cone 2 u
 102 
     
     
     
     "
`;

exports[`look.__cone 3`] = `
"__cone 3 u
 31024 
       
       
       
       
       
       "
`;

exports[`look.__cone 4`] = `
"__cone 4 u
 5310246 
         
         
         
         
         
         
         
         "
`;

exports[`look.__cone 5`] = `
"__cone 1 l
   
0  
   "
`;

exports[`look.__cone 6`] = `
"__cone 2 l
     
2    
0    
1    
     "
`;

exports[`look.__cone 7`] = `
"__cone 3 l
       
4      
2      
0      
1      
3      
       "
`;

exports[`look.__cone 8`] = `
"__cone 4 l
         
6        
4        
2        
0        
1        
3        
5        
         "
`;

exports[`look.__cone 9`] = `
"__cone 1 r
   
  0
   "
`;

exports[`look.__cone 10`] = `
"__cone 2 r
     
    1
    0
    2
     "
`;

exports[`look.__cone 11`] = `
"__cone 3 r
       
      3
      1
      0
      2
      4
       "
`;

exports[`look.__cone 12`] = `
"__cone 4 r
         
        5
        3
        1
        0
        2
        4
        6
         "
`;

exports[`look.__cone 13`] = `
"__cone 1 d
   
   
 0 "
`;

exports[`look.__cone 14`] = `
"__cone 2 d
     
     
     
     
 201 "
`;

exports[`look.__cone 15`] = `
"__cone 3 d
       
       
       
       
       
       
 42013 "
`;

exports[`look.__cone 16`] = `
"__cone 4 d
         
         
         
         
         
         
         
         
 6420135 "
`;

exports[`look.__f 1`] = `
"__f 1 u
 0 
   
   "
`;

exports[`look.__f 2`] = `
"__f 2 u
  0  
     
     
     
     "
`;

exports[`look.__f 3`] = `
"__f 3 u
   0   
       
       
       
       
       
       "
`;

exports[`look.__f 4`] = `
"__f 4 u
    0    
         
         
         
         
         
         
         
         "
`;

exports[`look.__f 5`] = `
"__f 1 l
   
0  
   "
`;

exports[`look.__f 6`] = `
"__f 2 l
     
     
0    
     
     "
`;

exports[`look.__f 7`] = `
"__f 3 l
       
       
       
0      
       
       
       "
`;

exports[`look.__f 8`] = `
"__f 4 l
         
         
         
         
0        
         
         
         
         "
`;

exports[`look.__f 9`] = `
"__f 1 r
   
  0
   "
`;

exports[`look.__f 10`] = `
"__f 2 r
     
     
    0
     
     "
`;

exports[`look.__f 11`] = `
"__f 3 r
       
       
       
      0
       
       
       "
`;

exports[`look.__f 12`] = `
"__f 4 r
         
         
         
         
        0
         
         
         
         "
`;

exports[`look.__f 13`] = `
"__f 1 d
   
   
 0 "
`;

exports[`look.__f 14`] = `
"__f 2 d
     
     
     
     
  0  "
`;

exports[`look.__f 15`] = `
"__f 3 d
       
       
       
       
       
       
   0   "
`;

exports[`look.__f 16`] = `
"__f 4 d
         
         
         
         
         
         
         
         
    0    "
`;

exports[`look.__three 1`] = `
"__three 1 u
102
   
   "
`;

exports[`look.__three 2`] = `
"__three 2 u
 102 
     
     
     
     "
`;

exports[`look.__three 3`] = `
"__three 3 u
  102  
       
       
       
       
       
       "
`;

exports[`look.__three 4`] = `
"__three 4 u
   102   
         
         
         
         
         
         
         
         "
`;

exports[`look.__three 5`] = `
"__three 1 l
2  
0  
1  "
`;

exports[`look.__three 6`] = `
"__three 2 l
     
2    
0    
1    
     "
`;

exports[`look.__three 7`] = `
"__three 3 l
       
       
2      
0      
1      
       
       "
`;

exports[`look.__three 8`] = `
"__three 4 l
         
         
         
2        
0        
1        
         
         
         "
`;

exports[`look.__three 9`] = `
"__three 1 r
  1
  0
  2"
`;

exports[`look.__three 10`] = `
"__three 2 r
     
    1
    0
    2
     "
`;

exports[`look.__three 11`] = `
"__three 3 r
       
       
      1
      0
      2
       
       "
`;

exports[`look.__three 12`] = `
"__three 4 r
         
         
         
        1
        0
        2
         
         
         "
`;

exports[`look.__three 13`] = `
"__three 1 d
   
   
201"
`;

exports[`look.__three 14`] = `
"__three 2 d
     
     
     
     
 201 "
`;

exports[`look.__three 15`] = `
"__three 3 d
       
       
       
       
       
       
  201  "
`;

exports[`look.__three 16`] = `
"__three 4 d
         
         
         
         
         
         
         
         
   201   "
`;

exports[`look.box 1`] = `
"box 1 u
012
3 4
567"
`;

exports[`look.box 2`] = `
"box 2 u
89abc
d012e
f3 4g
h567i
jklmn"
`;

exports[`look.box 3`] = `
"box 3 u
opqrstu
v89abcw
xd012ey
zf3 4gA
Bh567iC
DjklmnE
FGHIJKL"
`;

exports[`look.box 4`] = `
"box 4 u
MNOPQRSTU
VopqrstuW
Xv89abcwY
Zxd012ey!
!zf3 4gA!
!Bh567iC!
!DjklmnE!
!FGHIJKL!
!!!!!!!!!"
`;

exports[`look.box 5`] = `
"box 1 l
247
1 6
035"
`;

exports[`look.box 6`] = `
"box 2 l
cegin
b247m
a1 6l
9035k
8dfhj"
`;

exports[`look.box 7`] = `
"box 3 l
uwyACEL
tceginK
sb247mJ
ra1 6lI
q9035kH
p8dfhjG
ovxzBDF"
`;

exports[`look.box 8`] = `
"box 4 l
UWY!!!!!!
TuwyACEL!
StceginK!
Rsb247mJ!
Qra1 6lI!
Pq9035kH!
Op8dfhjG!
NovxzBDF!
MVXZ!!!!!"
`;

exports[`look.box 9`] = `
"box 1 r
530
6 1
742"
`;

exports[`look.box 10`] = `
"box 2 r
jhfd8
k5309
l6 1a
m742b
nigec"
`;

exports[`look.box 11`] = `
"box 3 r
FDBzxvo
Gjhfd8p
Hk5309q
Il6 1ar
Jm742bs
Knigect
LECAywu"
`;

exports[`look.box 12`] = `
"box 4 r
!!!!!ZXVM
!FDBzxvoN
!Gjhfd8pO
!Hk5309qP
!Il6 1arQ
!Jm742bsR
!KnigectS
!LECAywuT
!!!!!!YWU"
`;

exports[`look.box 13`] = `
"box 1 d
765
4 3
210"
`;

exports[`look.box 14`] = `
"box 2 d
nmlkj
i765h
g4 3f
e210d
cba98"
`;

exports[`look.box 15`] = `
"box 3 d
LKJIHGF
EnmlkjD
Ci765hB
Ag4 3fz
ye210dx
wcba98v
utsrqpo"
`;

exports[`look.box 16`] = `
"box 4 d
!!!!!!!!!
!LKJIHGF!
!EnmlkjD!
!Ci765hB!
!Ag4 3fz!
!ye210dxZ
Ywcba98vX
WutsrqpoV
UTSRQPONM"
`;

exports[`look.circle 1`] = `
"circle 1 u
 0 
2 1
 3 "
`;

exports[`look.circle 2`] = `
"circle 2 u
  4  
 605 
82 17
 a39 
  b  "
`;

exports[`look.circle 3`] = `
"circle 3 u
   c   
  e4d  
 g605f 
i82 17h
 ma39l 
  kbj  
   n   "
`;

exports[`look.circle 4`] = `
"circle 4 u
    o    
   qcp   
  se4dr  
 ug605ft 
wi82 17hv
 Cma39lB 
  Akbjz  
   ynx   
    D    "
`;

exports[`look.circle 5`] = `
"circle 1 l
 1 
0 3
 2 "
`;

exports[`look.circle 6`] = `
"circle 2 l
  7  
 519 
40 3b
 62a 
  8  "
`;

exports[`look.circle 7`] = `
"circle 3 l
   h   
  f7l  
 d519j 
c40 3bn
 e62ak 
  g8m  
   i   "
`;

exports[`look.circle 8`] = `
"circle 4 l
    v    
   thB   
  rf7lz  
 pd519jx 
oc40 3bnD
 qe62aky 
  sg8mA  
   uiC   
    w    "
`;

exports[`look.circle 9`] = `
"circle 1 r
 2 
3 0
 1 "
`;

exports[`look.circle 10`] = `
"circle 2 r
  8  
 a26 
b3 04
 915 
  7  "
`;

exports[`look.circle 11`] = `
"circle 3 r
   i   
  m8g  
 ka26e 
nb3 04c
 j915d 
  l7f  
   h   "
`;

exports[`look.circle 12`] = `
"circle 4 r
    w    
   Ciu   
  Am8gs  
 yka26eq 
Dnb3 04co
 xj915dp 
  zl7fr  
   Bht   
    v    "
`;

exports[`look.circle 13`] = `
"circle 1 d
 3 
1 2
 0 "
`;

exports[`look.circle 14`] = `
"circle 2 d
  b  
 93a 
71 28
 506 
  4  "
`;

exports[`look.circle 15`] = `
"circle 3 d
   n   
  jbk  
 l93am 
h71 28i
 f506g 
  d4e  
   c   "
`;

exports[`look.circle 16`] = `
"circle 4 d
    D    
   xny   
  zjbkA  
 Bl93amC 
vh71 28iw
 tf506gu 
  rd4es  
   pcq   
    o    "
`;

exports[`look.cone 1`] = `
"cone 1 u
 0 
   
   "
`;

exports[`look.cone 2`] = `
"cone 2 u
 213 
  0  
     
     
     "
`;

exports[`look.cone 3`] = `
"cone 3 u
 75468 
  213  
   0   
       
       
       
       "
`;

exports[`look.cone 4`] = `
"cone 4 u
 eca9bdf 
  75468  
   213   
    0    
         
         
         
         
         "
`;

exports[`look.cone 5`] = `
"cone 1 l
   
0  
   "
`;

exports[`look.cone 6`] = `
"cone 2 l
     
3    
10   
2    
     "
`;

exports[`look.cone 7`] = `
"cone 3 l
       
8      
63     
410    
52     
7      
       "
`;

exports[`look.cone 8`] = `
"cone 4 l
         
f        
d8       
b63      
9410     
a52      
c7       
e        
         "
`;

exports[`look.cone 9`] = `
"cone 1 r
   
  0
   "
`;

exports[`look.cone 10`] = `
"cone 2 r
     
    2
   01
    3
     "
`;

exports[`look.cone 11`] = `
"cone 3 r
       
      7
     25
    014
     36
      8
       "
`;

exports[`look.cone 12`] = `
"cone 4 r
         
        e
       7c
      25a
     0149
      36b
       8d
        f
         "
`;

exports[`look.cone 13`] = `
"cone 1 d
   
   
 0 "
`;

exports[`look.cone 14`] = `
"cone 2 d
     
     
     
  0  
 312 "
`;

exports[`look.cone 15`] = `
"cone 3 d
       
       
       
       
   0   
  312  
 86457 "
`;

exports[`look.cone 16`] = `
"cone 4 d
         
         
         
         
         
    0    
   312   
  86457  
 fdb9ace "
`;

exports[`look.f 1`] = `
"f 1 u
 0 
   
   "
`;

exports[`look.f 2`] = `
"f 2 u
  1  
  0  
     
     
     "
`;

exports[`look.f 3`] = `
"f 3 u
   2   
   1   
   0   
       
       
       
       "
`;

exports[`look.f 4`] = `
"f 4 u
    3    
    2    
    1    
    0    
         
         
         
         
         "
`;

exports[`look.f 5`] = `
"f 1 l
   
0  
   "
`;

exports[`look.f 6`] = `
"f 2 l
     
     
10   
     
     "
`;

exports[`look.f 7`] = `
"f 3 l
       
       
       
210    
       
       
       "
`;

exports[`look.f 8`] = `
"f 4 l
         
         
         
         
3210     
         
         
         
         "
`;

exports[`look.f 9`] = `
"f 1 r
   
  0
   "
`;

exports[`look.f 10`] = `
"f 2 r
     
     
   01
     
     "
`;

exports[`look.f 11`] = `
"f 3 r
       
       
       
    012
       
       
       "
`;

exports[`look.f 12`] = `
"f 4 r
         
         
         
         
     0123
         
         
         
         "
`;

exports[`look.f 13`] = `
"f 1 d
   
   
 0 "
`;

exports[`look.f 14`] = `
"f 2 d
     
     
     
  0  
  1  "
`;

exports[`look.f 15`] = `
"f 3 d
       
       
       
       
   0   
   1   
   2   "
`;

exports[`look.f 16`] = `
"f 4 d
         
         
         
         
         
    0    
    1    
    2    
    3    "
`;

exports[`look.three 1`] = `
"three 1 u
102
   
   "
`;

exports[`look.three 2`] = `
"three 2 u
 435 
 102 
     
     
     "
`;

exports[`look.three 3`] = `
"three 3 u
  768  
  435  
  102  
       
       
       
       "
`;

exports[`look.three 4`] = `
"three 4 u
   a9b   
   768   
   435   
   102   
         
         
         
         
         "
`;

exports[`look.three 5`] = `
"three 1 l
2  
0  
1  "
`;

exports[`look.three 6`] = `
"three 2 l
     
52   
30   
41   
     "
`;

exports[`look.three 7`] = `
"three 3 l
       
       
852    
630    
741    
       
       "
`;

exports[`look.three 8`] = `
"three 4 l
         
         
         
b852     
9630     
a741     
         
         
         "
`;

exports[`look.three 9`] = `
"three 1 r
  1
  0
  2"
`;

exports[`look.three 10`] = `
"three 2 r
     
   14
   03
   25
     "
`;

exports[`look.three 11`] = `
"three 3 r
       
       
    147
    036
    258
       
       "
`;

exports[`look.three 12`] = `
"three 4 r
         
         
         
     147a
     0369
     258b
         
         
         "
`;

exports[`look.three 13`] = `
"three 1 d
   
   
201"
`;

exports[`look.three 14`] = `
"three 2 d
     
     
     
 201 
 534 "
`;

exports[`look.three 15`] = `
"three 3 d
       
       
       
       
  201  
  534  
  867  "
`;

exports[`look.three 16`] = `
"three 4 d
         
         
         
         
         
   201   
   534   
   867   
   b9a   "
`;

exports[`xy2index 1`] = `
"     
 6   
  1  
 4 3 
5   2"
`;
